File: sources/Android/android-communications/src/main/java/polar/com/sdk/api/PolarBleApiCallbackProvider.java
// Copyright © 2019 Polar Electro Oy. All rights reserved.
package polar.com.sdk.api;
import androidx.annotation.NonNull;

import java.util.UUID;

import polar.com.sdk.api.model.PolarDeviceInfo;
import polar.com.sdk.api.model.PolarHrData;

/**
 * Contains the callbacks of the API.
 */
public interface PolarBleApiCallbackProvider {

    /**
     * @param   powered true = Bluetooth power on, false = Bluetooth power off
     */
    void blePowerStateChanged(final boolean powered);

    /**
     * Device is now connected
     * @param polarDeviceInfo Polar device information
     */
    void deviceConnected(@NonNull final PolarDeviceInfo polarDeviceInfo);

    /**
     * Connecting to device
     * @param polarDeviceInfo Polar device information
     */
    void deviceConnecting(@NonNull final PolarDeviceInfo polarDeviceInfo);

    /**
     * Device is now disconnected, no further action is needed from the application
     * if polar.com.sdk.api.PolarBleApi#disconnectFromPolarDevice is not called. Device will be automatically reconnected
     * @param polarDeviceInfo Polar device information
     */
    void deviceDisconnected(@NonNull final PolarDeviceInfo polarDeviceInfo);

    /**
     * Polar device's ECG feature is ready. Application may start ECG stream now if desired.
     * requires feature PolarBleApi#FEATURE_POLAR_SENSOR_STREAMING
     * @param   identifier Polar device id
     */
    void ecgFeatureReady(@NonNull final String identifier);

    /**
     * Polar device's ACC feature is ready. Application may start ACC stream now if desired.
     * requires feature PolarBleApi#FEATURE_POLAR_SENSOR_STREAMING
     * @param   identifier Polar device id
     */
    void accelerometerFeatureReady(@NonNull final String identifier);

    /**
     * Polar device's PPG feature is ready. Application may start PPG stream now if desired.
     * requires feature PolarBleApi#FEATURE_POLAR_SENSOR_STREAMING
     * @param   identifier Polar device id
     */
    void ppgFeatureReady(@NonNull final String identifier);

    /**
     * Polar device's PPI feature is ready. Application may start PPI stream now if desired.
     * requires feature PolarBleApi#FEATURE_POLAR_SENSOR_STREAMING
     * @param   identifier Polar device id
     */
    void ppiFeatureReady(@NonNull final String identifier);

    /**
     * Polar device's bioz feature is ready.
     * @param   identifier Polar device id
     */
    void biozFeatureReady(@NonNull final String identifier);

    /**
     * Polar device HR client is now ready and HR transmission is starting in a moment.
     * @param   identifier Polar device id or bt address
     */
    void hrFeatureReady(@NonNull final String identifier);

    /**
     * DIS information received
     * requires feature PolarBleApi#FEATURE_DEVICE_INFO
     * @param   identifier Polar device id or bt address
     * @param   uuid uuid of dis value
     * @param   value dis value for uuid
     */
    void disInformationReceived(@NonNull final String identifier, @NonNull UUID uuid, @NonNull final String value);

    /**
     * Battery level received
     * requires feature PolarBleApi#FEATURE_BATTERY_INFO
     * @param   identifier Polar device id or bt address
     * @param   level battery level (value between 0-100%)
     */
    void batteryLevelReceived(@NonNull final String identifier, final int level);

    /**
     * HR notification data received from device. Notice when using OH1
     * and PPI measurement is started hr received from this callback is 0.
     * @param   identifier Polar device id or bt address
     * @param   data @see polar.com.sdk.api.model.PolarHrData.java
     */
    void hrNotificationReceived(@NonNull final String identifier,@NonNull final PolarHrData data);

    /**
     * File transfer ready
     * requires feature PolarBleApi#FEATURE_POLAR_FILE_TRANSFER
     * @param   identifier Polar device id
     */
    void polarFtpFeatureReady(@NonNull final String identifier);
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarBleApiCallbackProvider.kt
// Copyright © 2019 Polar Electro Oy. All rights reserved.
package com.polar.sdk.api

import com.polar.androidcommunications.api.ble.model.DisInfo
import com.polar.androidcommunications.api.ble.model.gatt.client.ChargeState
import com.polar.androidcommunications.api.ble.model.gatt.client.PowerSourcesState
import com.polar.sdk.api.model.PolarDeviceInfo
import com.polar.sdk.api.model.PolarHealthThermometerData
import com.polar.sdk.api.model.PolarHrData
import java.util.*

/**
 * Contains the callbacks of the API.
 */
interface PolarBleApiCallbackProvider {
    /**
     * Bluetooth power state of the device where this SDK is running
     *
     * @param powered true = Bluetooth power on, false = Bluetooth power off
     */
    fun blePowerStateChanged(powered: Boolean)

    /**
     * Device is now connected
     *
     * @param polarDeviceInfo Polar device information
     */
    fun deviceConnected(polarDeviceInfo: PolarDeviceInfo)

    /**
     * Connecting to device
     *
     * @param polarDeviceInfo Polar device information
     */
    fun deviceConnecting(polarDeviceInfo: PolarDeviceInfo)

    /**
     * Device is now disconnected
     *
     * @param polarDeviceInfo Polar device information
     */
    fun deviceDisconnected(polarDeviceInfo: PolarDeviceInfo)

    /**
     * The feature is available in this device and it is ready. Called only for the features which are specified in [PolarBleApi] construction.
     *
     * @param identifier Polar device id
     * @param feature   feature is ready
     */
    fun bleSdkFeatureReady(identifier: String, feature: PolarBleApi.PolarBleSdkFeature)

    /**
     * Called once all requested features have been evaluated after a device connection.
     * This is the simplest way to wait for the device to become fully ready for further operations.
     *
     * Constructing [PolarBleApi] with an empty feature set causes the SDK to evaluate all SDK features by default.
     * SDK calls this method after a device is connected.
     *
     * The [ready] list contains features that are confirmed ready for use.
     * The [unavailable] list contains features that are not supported by this device.
     * Features absent from both lists timed out before their readiness could be established;
     * they can still become ready later and will be reported via [bleSdkFeatureReady].
     *
     * @param identifier Polar device id
     * @param ready features that are confirmed ready for use on this device
     * @param unavailable features that are not supported by this device
     */
    fun bleSdkFeaturesReadiness(identifier: String, ready: List<PolarBleApi.PolarBleSdkFeature>, unavailable: List<PolarBleApi.PolarBleSdkFeature>)

    /**
     * DIS information received. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_DEVICE_INFO]
     *
     * @param identifier Polar device id or bt address
     * @param uuid       uuid of dis value
     * @param value      dis value for uuid
     */
    fun disInformationReceived(identifier: String, uuid: UUID, value: String)

    /**
     * DIS information received. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_DEVICE_INFO]
     *
     * @param identifier Polar device id or bt address
     * @param disInfo    [DisInfo] key-value pair
     */
    fun disInformationReceived(identifier: String, disInfo: DisInfo)

    /**
     * Battery level received. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_BATTERY_INFO]
     *
     * @param identifier Polar device id or bt address
     * @param level      battery level (value between 0-100%)
     */
    fun batteryLevelReceived(identifier: String, level: Int)

    /**
     * Battery charging status received. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_BATTERY_INFO]
     *
     * @param identifier Polar device id or bt address
     * @param chargingStatus Charging status [ChargeState]
     */
    fun batteryChargingStatusReceived(identifier: String, chargingStatus: ChargeState)

    /**
     * Battery power sources state received. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_BATTERY_INFO]
     *
     * @param identifier Polar device id or bt address
     * @param powerSourcesState [PowerSourcesState]
     */
    fun powerSourcesStateReceived(identifier: String, powerSourcesState: PowerSourcesState)

    /**
     * HR notification data received from device. Notice when using OH1
     * and PPI measurement is started hr received from this callback is 0.
     *
     * @param identifier Polar device id or bt address
     * @param data       @see polar.com.sdk.api.model.PolarHrData.java
     */
    fun hrNotificationReceived(identifier: String, data: PolarHrData.PolarHrSample)

    /**
     * Health Thermometer Service. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_HTS]
     *
     * @param identifier Polar device id or bt address
     * @param data       @see polar.com.sdk.api.model.PolarHealthThermometerData.kt
     */
    fun htsNotificationReceived(identifier: String, data: PolarHealthThermometerData)
}